[go-backend] Logout API

the-spectator opened the pull request:
## Description
This PR implements Logout API for peerly.

Continuation #40

jahio commented on go-backend/db/user_blacklisted_tokens.go:
```diff
@@ -17,3 +30,13 @@ func (s *pgStore) CleanBlacklistedTokens() (err error) {
 }
+
+func (s *pgStore) CreateUserBlacklistedToken(ctx context.Context, blacklistedToken UserBlacklistedToken) (err error) {
+	_, err = s.db.Exec(insertBlacklistedToken, blacklistedToken.UserID, blacklistedToken.Token, blacklistedToken.ExpirationDate)
```
I would suggest using `s.db.MustExec(...)` here instead so we'll be sure to get a failure if that query doesn't execute for some reason. If we're blacklisting a token, we want to make **sure** it no longer has access to do anything, and if that process fails we need it to throw an error so we can get alerted and do something sooner rather than later.

the-spectator replied:
The downside of `s.db.MustExec` is if it fails it `panics` and the request is dropped, this is not will not allow us to further process request. i.e we will not be able to return our error response to client. Hence I would prefer `db.Exec` here.

jahio commented on go-backend/db/user_blacklisted_tokens.go:
```diff
@@ -17,3 +30,13 @@ func (s *pgStore) CleanBlacklistedTokens() (err error) {
+	_, err = s.db.Exec(insertBlacklistedToken, blacklistedToken.UserID, blacklistedToken.Token, blacklistedToken.ExpirationDate)
+
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error inserting the blacklisted token")
```
Suggestion:

```go
logger.WithField("err", err.Error()).Error("Error inserting the blacklisted token for user with id " + blacklistedToken.UserID)
```

So our errors - and whatever ends up reporting them - have something useful to duplicate the problem in the log file. (It's ok if this is done in a future commit, imo.)

the-spectator replied:
Implemented in  fb35e6c

jahio commented on go-backend/service/session_http.go:
```diff
@@ -0,0 +1,75 @@
+	newTkn, err = jwt.ParseWithClaims(token, claims, func(token *jwt.Token) (interface{}, error) {
+		return config.JwtKey(), nil
+	})
+
```
I recommend checking for errors here:

```go
if err != nil {
  // do something
}
```

the-spectator replied:
Resolved in fb35e6c

jahio commented on go-backend/service/session_http.go:
```diff
@@ -0,0 +1,75 @@
+			} else if err == errInvalidToken {
+				rw.WriteHeader(http.StatusUnauthorized)
+				return
+			}
```
I recommend adding an additional `else` block here as a catch-all, with a `http.StatusUnauthorized` header being written out. Maybe we can later add some conditional logging (e.g. "if verbose mode, log user id and (maybe?) info about the token")

the-spectator replied:
Good Idea, I am thinking of using `rw.WriteHeader(http.StatusBadRequest)` for catch-all. as status `Bad Request` feels good fit here. Implemented in fb35e6c

jahio commented on go-backend/service/session_http.go:
```diff
@@ -0,0 +1,75 @@
+
+		err = deps.Store.CreateUserBlacklistedToken(req.Context(), userBlackListedToken)
+		if err != nil {
+			rw.WriteHeader(http.StatusInternalServerError)
```
Maybe we can write out a simple JSON response here, something like:

```json
{ "status":500, "error":"Internal server error inserting token into blacklisted tokens table" }
```

the-spectator replied:
Nice Idea. Implemented in fb35e6cb9068eb499940d166aed7a0ea1a512736

jahio commented on go-backend/db/user.go:
```diff
@@ -2,13 +2,24 @@ package db
-	Age  int    `db:"age" json:"age"`
+	ID              int64     `db:"id" json:"id"`
+	Name            string    `db:"name" json:"full_name"`
+	Age             int       `db:"age" json:"age"`
```
We need to remove `Age` here. We don't need to track it in the application and may have legal/hr implications in some areas.

the-spectator replied:
Nice catch! it was extra attribute not being used in user schema. Removed in 60298da0af68584d9c932aa3d2b361cbf8cb0f40

jahio commented:
Looks pretty good Akshay! There are a few areas where we could improve error handling and messaging but those are minor concerns. I went ahead and commented on those but I'll leave it up to you if you want to implement those things now or address them later - like I said, they're pretty minor. 

The important thing is that this branch _builds_ 😃

<img width="548" alt="image" src="https://user-images.githubusercontent.com/421850/80111805-98b94780-853d-11ea-9dd2-fb0daef8c54c.png">

I'm ok with merging it as-is, but would like you to have a chance to comment/discuss the points I made above. If you don't agree with them that's totally fine, we'll merge it anyway. (This is just to explain the only reason I'm not merging it right now.)

Feel free to merge it in yourself at your own convenience, or once I hear from you that you think we should merge it now, I'll do it. Either way is fine by me.

the-spectator commented:
@jahio Thank you for your review, I have addressed most of the review comments but chose to prefer `db.Exec` over `db.MustExec` as per the reasons stated in https://github.com/joshsoftware/peerly/pull/63#discussion_r414120984
